cmake/introspection.cmake

```diff
@@ -15,3 +17,243 @@ check_include_file_cxx(sys/prctl.h HAVE_SYS_PRCTL_H)
+    if(NOT IFADDR_LINKS_WITHOUT_LIBSOCKET)
+      check_cxx_source_links_with_libs(socket "${check_socket_source}" IFADDR_NEEDS_LINK_TO_LIBSOCKET)
+      if(IFADDR_NEEDS_LINK_TO_LIBSOCKET)
+        link_libraries(socket)
+      else()
+        message(FATAL_ERROR "Cannot figure out how to use getifaddrs/freeifaddrs.")
+      endif()
+    endif()
+  endif()
+endif()
+
+# Check for gmtime_r(), fallback to gmtime_s() if that is unavailable.
+# Fail if neither are available.
+check_cxx_source_compiles("
+  #include <ctime>
+
+  int main()
+  {
+    gmtime_r((const time_t*)nullptr, (struct tm*)nullptr);
+  }
+  " HAVE_GMTIME_R
+)
+if(NOT HAVE_GMTIME_R)
+  check_cxx_source_compiles("
+    #include <ctime>
+
+    int main()
+    {
+      gmtime_s((struct tm*)nullptr, (const time_t*)nullptr);
+    }
+    " HAVE_GMTIME_S
+  )
+  if(NOT HAVE_GMTIME_S)
+    message(FATAL_ERROR "Both gmtime_r and gmtime_s are unavailable.")
+  endif()
+endif()
+
+check_cxx_source_compiles("
+  #include <cstdlib>
+
+  int main()
+  {
+    return std::system(nullptr);
+  }
+  " HAVE_STD_SYSTEM
+)
+check_cxx_source_compiles("
+  #include <stdlib.h>
+
+  int main()
+  {
+    return ::_wsystem(NULL);
+  }
+  " HAVE__WSYSTEM
+)
+if(HAVE_STD_SYSTEM OR HAVE__WSYSTEM)
+  set(HAVE_SYSTEM 1)
+endif()
+
+check_include_file_cxx(string.h HAVE_STRING_H)
+if(HAVE_STRING_H)
+  check_cxx_source_compiles("
+    #include <string.h>
+
+    int main()
+    {
+      char buf[100];
+      char x = *strerror_r(0, buf, sizeof buf);
+      char* p = strerror_r(0, buf, sizeof buf);
+      return !p || x;
+    }
+    " STRERROR_R_CHAR_P
+  )
+endif()
+
+# Check for malloc_info (for memory statistics information in getmemoryinfo).
+check_cxx_symbol_exists(malloc_info "malloc.h" HAVE_MALLOC_INFO)
+
+# Check for mallopt(M_ARENA_MAX) (to set glibc arenas).
+check_cxx_source_compiles("
+  #include <malloc.h>
+
+  int main()
+  {
+    mallopt(M_ARENA_MAX, 1);
+  }
+  " HAVE_MALLOPT_ARENA_MAX
+)
+
+# Check for posix_fallocate().
+check_cxx_source_compiles("
+  // same as in src/util/system.cpp
+  #ifdef __linux__
+  #ifdef _POSIX_C_SOURCE
+  #undef _POSIX_C_SOURCE
+  #endif
+  #define _POSIX_C_SOURCE 200112L
+  #endif // __linux__
+  #include <fcntl.h>
+
+  int main()
+  {
+    return posix_fallocate(0, 0, 0);
+  }
+  " HAVE_POSIX_FALLOCATE
+)
+
+# Check for strong getauxval() support in the system headers.
+check_cxx_source_compiles("
+  #include <sys/auxv.h>
+
+  int main()
+  {
+    getauxval(AT_HWCAP);
+  }
+  " HAVE_STRONG_GETAUXVAL
+)
+
+# Check for different ways of gathering OS randomness:
+# - Linux getrandom()
+check_cxx_source_compiles("
+  #include <unistd.h>
+  #include <sys/syscall.h>
+  #include <linux/random.h>
+
+  int main()
+  {
+    syscall(SYS_getrandom, nullptr, 32, 0);
+  }
+  " HAVE_SYS_GETRANDOM
+)
+
+# - BSD getentropy()
+check_cxx_symbol_exists(getentropy "unistd.h;sys/random.h" HAVE_GETENTROPY_RAND)
```

Comment:
We are sometimes using `check_cxx_source_compiles`, like with `SYS_getrandom` just above, and sometimes `check_cxx_symbols_exists` like here, while both `SYS_getrandom` and `getentropy` are checked with `AC_COMPILE_IFELSE` in autotools currently. What is logic that defines when either of the checks are used?

Reply:
The [`check_cxx_symbol_exists`](https://cmake.org/cmake/help/latest/module/CheckCXXSymbolExists.html) command is intended to be used in simple cases.

Talking about the `SYS_getrandom`, it is called indirectly, and a more general `check_cxx_source_compiles()` command seems more appropriate here.

Reply:
Ok, I see that these criteria make sense here, e.g. we use `check_cxx_symbol_exists` for symbols with simple types called directly. Would it then make sense to use `check_cxx_symbol_exists` for `std::system` and `::_wsystem`?

Reply:
> Would it then make sense to use `check_cxx_symbol_exists` for `std::system` and `::_wsystem`?

Thanks! [Updated](https://github.com/hebasto/bitcoin/pull/6#issuecomment-1442023199).